File: packages/core/src/programmers/helpers/HttpMetadataUtil.ts
import { MetadataSchema } from "../../schemas/metadata/MetadataSchema";

export namespace HttpMetadataUtil {
  export const atomics = (
    metadata: MetadataSchema,
  ): Set<"boolean" | "bigint" | "number" | "string"> =>
    new Set([
      ...metadata.atomics.map((a) => a.type),
      ...metadata.constants.map((c) => c.type),
      ...(metadata.templates.length ? (["string"] as const) : []),
    ]);

  export const isUnion = (metadata: MetadataSchema): boolean =>
    atomics(metadata).size +
      metadata.arrays.length +
      metadata.tuples.length +
      metadata.natives.length +
      metadata.maps.length +
      metadata.objects.length >
    1;
}

File: packages/core/src/programmers/helpers/HttpMetadataUtil.go
package helpers

type httpMetadataUtilNamespace struct{}

var HttpMetadataUtil = httpMetadataUtilNamespace{}

func (httpMetadataUtilNamespace) Atomics(metadata Metadata) []string {
	set := map[string]bool{}
	for _, atomic := range GetSlice(metadata, "atomics") {
		if name := GetName(atomic); name != "" {
			set[name] = true
		}
	}
	for _, constant := range GetSlice(metadata, "constants") {
		if name := GetName(constant); name != "" {
			set[name] = true
		}
	}
	if len(GetSlice(metadata, "templates")) != 0 {
		set["string"] = true
	}
	out := make([]string, 0, len(set))
	for key := range set {
		out = append(out, key)
	}
	return out
}

func (httpMetadataUtilNamespace) IsUnion(metadata Metadata) bool {
	count := len(HttpMetadataUtil.Atomics(metadata))
	for _, key := range []string{"arrays", "tuples", "natives", "maps", "sets", "objects"} {
		count += len(GetSlice(metadata, key))
	}
	return count > 1
}
